server/keycloak.js

```diff
@@ -0,0 +1,121 @@
+    this.authUrl = process.env.KEYCLOAK_AUTH_URL;
+    this.clientNameFrontend = process.env.KEYCLOAK_FE_CLIENTID;
+    this.clientNameBackend = process.env.KEYCLOAK_API_CLIENTID;
+    this.clientSecretBackend = process.env.KEYCLOAK_API_SECRET;
+    this.serviceAccountUsername = process.env.KEYCLOAK_SA_USERNAME;
+    this.serviceAccountPassword = process.env.KEYCLOAK_SA_PASSWORD;
+    const config = {
+      ...defaults,
+      realm: this.realm,
+      'auth-server-url': this.authUrl,
+      resource: this.clientNameBackend,
+    };
+    this.keycloakConnect = new KeyCloakConnect({}, config);
+  }
+
+  RealmInfoFrontend() {
+    return {
+      realm: this.realm,
+      url: this.authUrl,
+      clientId: this.clientNameFrontend,
+    };
+  }
+
+  expressMiddleware(...args) { // Default connect middleware for Keycloak connect library
+    return this.keycloakConnect.middleware(...args);
+  }
+
+  allowRolesMiddleware(...roles) { // Connect middleware for limiting roles
+    const allowRoles = (token) => {
+      if (token.isExpired()) return false;
+      if (roles.length === 1 && roles[0] === '*') return true; // Allows any role
+      if (token.hasRole('superuser')) return true;
+      return roles.some((role) => token.hasRole(role));
+    };
+    return this.keycloakConnect.protect(allowRoles);
+  }
+
+  getUserInfoMiddleware() { // Connect middleware for adding HCAP user info to request object
+    return (req, res, next) => {
+      // Optional chaining would be great here once ESLint supports it *sigh*
+      const { content } = req.kauth.grant.access_token;
+      const { roles } = content.resource_access[this.clientNameBackend] || { roles: [] };
+      req.hcapUserInfo = {
+        name: content.name,
+        roles,
+        regions: roles.map((role) => regionMap[role]).filter((region) => region),
+        isSuperUser: roles.includes('superuser'),
+        isMoH: roles.includes('ministry_of_health'),
+      };
+      next();
+    };
+  }
+
+  async authenticateServiceAccount() {
+    const data = querystring.stringify({
+      grant_type: 'password',
+      client_id: this.clientNameBackend,
+      client_secret: this.clientSecretBackend,
+      username: this.serviceAccountUsername,
+      password: this.serviceAccountPassword,
+    });
+    const config = { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } };
+    const response = await axios.post(`${this.authUrl}/realms/${this.realm}/protocol/openid-connect/token`, data, config);
+    this.access_token = response.data.access_token;
+  }
+
+  async authenticateIfNeeded() {
+    // Race condition if token expires between this call and the desired authenticated call
+    const config = { headers: { Authorization: `Bearer ${this.access_token}` } };
+    try {
+      await axios.get(`${this.authUrl}/realms/${this.realm}/protocol/openid-connect/userinfo`, config);
+    } catch (error) {
+      await this.authenticateServiceAccount();
```

Comment:
As a suggestion, I recommend, do a little bit digging on error before re-authenticate, like check is it timeout / network issue / 401.